lib/util.c

```diff
@@ -111,6 +113,55 @@ duo_common_ini_handler(struct duo_config *cfg, const char *section,
             /* Make timeout milliseconds */
             cfg->https_timeout *= 1000;
         }
+    } else if (strcmp(name, "user_map_file") == 0) {
+        FILE *f = fopen(val, "r");
+        struct stat stbuf;
+        if (f == NULL) {
+            fprintf(stderr, "Unable to open duo user_map_file %s\n", val);
+            return (0);
+        }
+        if (fstat(fileno(f), &stbuf) != 0) {
+            fprintf(stderr, "Could not stat duo user_map_file %s\n", val);
+            return (1);
+        }
+        if (stbuf.st_mode & (S_IROTH | S_IWOTH)) {
+            duo_syslog(LOG_ERR, "duo user_map_file %s is world readable/writable, bailing\n", val);
+            return (1);
+        }
+        char line[USER_MAP_MAX];
+        int i = 0;
+        struct user_map* last = NULL;
+        while (fgets(line, USER_MAP_MAX - 1, f) != NULL) {
+            ++i;
+            char *space = strchr(line, ' ');
+            if (space == NULL) {
+                fprintf(stderr, "line %d malformed in %s\n", i, val);
+                return (0);
+            }
+            struct user_map *map_cell = malloc(sizeof(struct user_map));
+            if (map_cell == NULL) {
+                fprintf(stderr, "Out of memory parsing user map file\n");
+                return (0);
+            }
+            map_cell->next = NULL;
+            size_t space_offset = space - line;
+            strncpy(map_cell->from, line, space_offset);
+            map_cell->from[space_offset] = '\0';
+            char *newline = strchr(line + space_offset, '\n');
+            if (newline != NULL) {
+                *newline = '\0';
+            }
+            strncpy(map_cell->to, line + space_offset + 1, USER_MAP_MAX - space_offset - 1);
+            // The above will include the trailing NULK from fgets(3)
```

Comment:
fgets always returns a trailing NUL, regardless of whether it included a newline. So the trailing NUL of this buffer will either be the trailing NUL inserted by fgets for the last line of a file without a trailing newline, or will be the one we insert ourselves on line 152.